### PR TITLE
Create an IQN example aimed at reproducing the original paper and its evaluation protocol.

### PR DESCRIPTION
**Pull request by prabhatnagarajan**

### CODE REVIEW

**Comment by prabhatnagarajan** [comment]
Checklist:
- **Evaluation Frequency**
After every 1 million frames (250K environment timesteps), the agent is evaluated for 500K frames.  This totals 200 evaluations  (Source:  PDF page 7 under "best agent performance" - https://arxiv.org/pdf/1710.10044.pdf)
- **Evaluation phase** - 500K frames or 125K timesteps (Source:  PDF page 7 under "best agent performance" - https://arxiv.org/pdf/1710.10044.pdf)
-**Evaluation Episode Length**
- Each evaluation is capped at 108K frames (or 108K/4 = 27K timesteps) (https://arxiv.org/pdf/1707.06887.pdf erratum, personal correspondence)
-**Evaluation Episode Starts**
Each evaluation episode begins with a random number of no-ops (up to 30), number chosen uniformly randomly (Source:  PDF page 7 under "best agent performance" - https://arxiv.org/pdf/1710.10044.pdf)
-**Evaluation Episode Policy**
Epsilon greedy policy with `epsilon=0.001`
-**Evaluation Output*
The best average amongst the periodic evaluations done during training (source: personal correspondence), _as well as_ a reevaluation of this best policy for 30 episodes (only done for ChainerRL benchmarks)

File: examples/atari/iqn/train_iqn.py [hunk 1]
```diff
@@ -62,7 +62,7 @@ def main():
     parser.add_argument('--quantile-thresholds-N', type=int, default=64)
     parser.add_argument('--quantile-thresholds-N-prime', type=int, default=64)
     parser.add_argument('--quantile-thresholds-K', type=int, default=32)
-    parser.add_argument('--n-best-episodes', type=int, default=30)
```
Comment (muupan) [inline examples/atari/iqn/train_iqn.py]: What is the reason for this change?

**Comment by muupan** [comment]
Not reviewed code yet but the checklist looks good

File: examples/atari/iqn/train_iqn.py (hunk 1 above)
Comment (prabhatnagarajan) [reply examples/atari/iqn/train_iqn.py]: I changed it to 200. The Rainbow paper does 200 evaluations, and since this paper doesn't have a re-evaluation of the best network, I figured that we should do what Rainbow does, since that's the most "up-to-date" evaluation protocol.

**Comment by prabhatnagarajan** [comment]
Results:

**Asterix**: 464666.666667 vs. Reported 342,016
**BeamRider**: 35525.0 vs. Reported 42,776
**Breakout**: 738.0 vs. Reported 734
**Seaquest**: 31670.0 vs. Reported 30,140
**Pong**: 21.0 vs. Reported 21.0
**SpaceInvaders**: 43649.0 vs. Reported 28,888
**Qbert**: 25971.3414634 vs. Reported 25,750

**Comment by prabhatnagarajan** [comment]
Learning curves:

Asterix-IQN
BeamRid-IQN
Breakout-IQN
Qbert-IQN
Pong-IQN
Seaquest-IQN
SpaceInvad-IQN

File: examples/atari/iqn/README.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+| Stargunner | N/A| **74677**|
+| Tennis | N/A| **23.6**|
+| Time Pilot | N/A| **12236**|
+| Tutankham | N/A| **293**|
+| Up’n Down | N/A| **88148**|
+| Venture | N/A| **1318**|
+| Video Pinball | N/A| **698045**|
+| WizardOfWor | N/A| **31190**|
+| YarsRevenge | N/A| **28379**|
+| Zaxxon | N/A| **21772**|
```
Comment (muupan) [inline examples/atari/iqn/README.md]: How about adding results we already have?

File: examples/atari/iqn/train_iqn.py [hunk 3]
```diff
@@ -161,25 +163,45 @@ def phi(x):
...
             outdir=args.outdir,
-            save_best_so_far_agent=False,
+            save_best_so_far_agent=True,
             eval_env=eval_env,
         )
 
+        dir_of_best_network = os.path.join(args.outdir, "best")
+        agent.load(dir_of_best_network)
+
+        # run 30 evaluation episodes, each capped at 5 mins of play
```
Comment (muupan) [inline examples/atari/iqn/train_iqn.py]: s/30/200/
s/5 mins/30 mins/

File: examples/atari/iqn/train_iqn.py [hunk 4]
```diff
@@ -21,6 +21,7 @@
 from chainerrl import misc
 from chainerrl import replay_buffer
 from chainerrl.wrappers import atari_wrappers
+import json
```
Comment (muupan) [inline examples/atari/iqn/train_iqn.py]: In ChainerRL, import sentences are ordered as
- `__future__`
- `future` and `builtins`
- standard library in alphabetical order
- third-party (gym, chainer, numpy etc) in alphabetical order
- chainerrl in alphabetical order

So, can you move `json` to the standard library section?

File: examples/atari/iqn/README.md [hunk 5]
```diff
@@ -0,0 +1,91 @@
+# IQN
+This example trains a IQN agent, from the following paper: [Implicit Quantile Networks for Distributional Reinforcement Learning](https://arxiv.org/pdf/1806.06923.pdf).
```
Comment (muupan) [inline examples/atari/iqn/README.md]: s/a IQN/an IQN/ ?
Use abs page, not pdf.

File: chainerrl/experiments/evaluator.py [hunk 6]
```diff
@@ -344,6 +351,14 @@ def evaluate_and_update_max_score(self, t, episodes):
...
             if self.save_best_so_far_agent:
                 save_agent(self.agent, "best", self.outdir, self.logger)
+                if not os.path.exists(os.path.join(self.outdir, 'best')):
+                    try:
+                        os.makedirs(os.path.join(self.outdir, 'best'))
+                    except OSError as err:
+                        if err.errno != errno.EEXIST:
+                            raise
+                record_best_eval(os.path.join(self.outdir, 'best'),
+                                 t, self.max_score)
```
Comment (muupan) [inline chainerrl/experiments/evaluator.py]: - `best` directory corresponds to an internal state of the agent state, so it is not intended to contain other files.
- Why do we need to save `best/best.txt`? We can see the max evaluation score in `scores.txt`.

**Comment by muupan** [comment]
The learning curves you pasted are limited to 6M steps. Can you add curves with full 50M steps?

**Comment by muupan** [comment]
Breakout's curve does not seem to reach 738.0. Do these learning curves plots "mean" in scores.txt?  Did you apply any smoothing?
